File: cssp-web/src/app/shared/service/relatorio.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {environment} from "../../../environments/environment";
import {Observable} from "rxjs";
import {RelatorioEntreDatasModel} from "../../model/report/relatorio-entre-datas.model";

@Injectable({
  providedIn: 'root'
})
export class RelatorioService {

  constructor(private http: HttpClient) {
  }

  resourceUrl = environment.apiUrl + '/relatorio';

  balanceReportProductInStock(): Observable<any> {
    return this.http.get(`${this.resourceUrl}/relatorio-estoque-produtos`, {responseType: 'arraybuffer'});
  }

  pointOfOrderProductReport(): Observable<any> {
    return this.http.get(`${this.resourceUrl}/relatorio-ponto-encomenda-produtos`, {responseType: 'arraybuffer'});
  }

  reportMostConsumedBeersInAPeriod(report: RelatorioEntreDatasModel): Observable<any> {
    return this.http.get(
      `${this.resourceUrl}/relatorio-chopes-populares?dataInicio=${report.dataInicial}&dataFinal=${report.dataFinal}`,
      {responseType: 'arraybuffer'});
  }

  customerReportWithAmountPurchasedInPeriod(report: RelatorioEntreDatasModel): Observable<any> {
    return this.http.get(
      `${this.resourceUrl}/relatorio-total-compras-clientes?dataInicio=${report.dataInicial}&dataFinal=${report.dataFinal}`,
      {responseType: 'arraybuffer'});
  }
  expenseAndIncomeMovementReport(report: RelatorioEntreDatasModel): Observable<any> {
    return this.http.get(
      `${this.resourceUrl}//relatorio-receita-despesas?dataInicio=${report.dataInicial}&dataFinal=${report.dataFinal}`,
      {responseType: 'arraybuffer'});
  }

}

File: service/src/main/java/com/cssp/choperiaselfservice/service/RelatorioService.java
package com.cssp.choperiaselfservice.service;

import lombok.RequiredArgsConstructor;
import net.sf.jasperreports.engine.*;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Transactional
public class RelatorioService {
    private Connection getConnection() throws SQLException {
        String url = "jdbc:postgresql://localhost:5432/postgres";
        String username = "admin";
        String password = "admin";
        return DriverManager.getConnection(url, username, password);
    }

    public byte[] balanceReportProductInStock() throws JRException, SQLException {
        Connection connection = getConnection();
        JasperReport jasperReport = JasperCompileManager.compileReport(
                getClass().getResourceAsStream("/report/relatorio-estoque-produtos.jrxml")
        );

        Map<String, Object> params = new HashMap<>();
        params.put(JRParameter.REPORT_CONNECTION, connection);

        JasperPrint jasperPrint = JasperFillManager.fillReport(jasperReport, params, connection);
        connection.close();
        return JasperExportManager.exportReportToPdf(jasperPrint);
    }

    public byte[] pointOfOrderProductReport() throws JRException, SQLException {
        Connection connection = getConnection();
        JasperReport jasperReport = JasperCompileManager.compileReport(
                getClass().getResourceAsStream("/report/relatorio-ponto-encomenda-produtos.jrxml")
        );

        Map<String, Object> params = new HashMap<>();
        params.put(JRParameter.REPORT_CONNECTION, connection);

        JasperPrint jasperPrint = JasperFillManager.fillReport(jasperReport, params, connection);
        connection.close();
        return JasperExportManager.exportReportToPdf(jasperPrint);
    }

    public byte[] reportMostConsumedBeersInAPeriod(Date dataInicial, Date dataFinal) throws JRException, SQLException {
        Connection connection = getConnection();
        JasperReport jasperReport = JasperCompileManager.compileReport(
                getClass().getResourceAsStream("/report/relatorio-chopes-populares.jrxml")
        );

        Map<String, Object> params = new HashMap<>();
        params.put(JRParameter.REPORT_CONNECTION, connection);
        params.put("DataInicial", dataInicial);
        params.put("DataFinal", dataFinal);

        JasperPrint jasperPrint = JasperFillManager.fillReport(jasperReport, params, connection);
        connection.close();
        return JasperExportManager.exportReportToPdf(jasperPrint);
    }

    public byte[] customerReportWithAmountPurchasedInPeriod(Date dataInicial, Date dataFinal) throws JRException, SQLException {
        Connection connection = getConnection();
        JasperReport jasperReport = JasperCompileManager.compileReport(
                getClass().getResourceAsStream("/report/relatorio-total-compras-clientes.jrxml")
        );

        Map<String, Object> params = new HashMap<>();
        params.put(JRParameter.REPORT_CONNECTION, connection);
        params.put("DataInicial", dataInicial);
        params.put("DataFinal", dataFinal);

        JasperPrint jasperPrint = JasperFillManager.fillReport(jasperReport, params, connection);
        connection.close();
        return JasperExportManager.exportReportToPdf(jasperPrint);
    }

    public byte[] expenseAndIncomeMovementReport(Date dataInicial, Date dataFinal) throws JRException, SQLException {
        Connection connection = getConnection();
        JasperReport jasperReport = JasperCompileManager.compileReport(
                getClass().getResourceAsStream("/report/relatorio-receita-despesas.jrxml")
        );

        Map<String, Object> params = new HashMap<>();
        params.put(JRParameter.REPORT_CONNECTION, connection);
        params.put("DataInicial", dataInicial);
        params.put("DataFinal", dataFinal);

        JasperPrint jasperPrint = JasperFillManager.fillReport(jasperReport, params, connection);
        connection.close();
        return JasperExportManager.exportReportToPdf(jasperPrint);
    }
}
